go/test/endtoend/vtgate/buffer/buffer_test.go

```diff
@@ -0,0 +1,453 @@
+	master.VttabletProcess.QueryTablet(query, keyspaceUnshardedName, true)
+	disableSemiSyncMasterQuery := "SET GLOBAL rpl_semi_sync_master_enabled = 0"
+	if master.VttabletProcess.EnableSemiSync {
+		master.VttabletProcess.QueryTablet(disableSemiSyncMasterQuery, keyspaceUnshardedName, true)
+	}
+
+	// Wait for replica to catch up to master.
+	waitForReplicationPos(t, &master, &replica, 60.0)
+
+	duration := time.Since(start)
+	minUnavailabilityInS := 1.0
+	if duration.Seconds() < minUnavailabilityInS {
+		w := minUnavailabilityInS - duration.Seconds()
+		fmt.Printf("Waiting for %.1f seconds because the failover was too fast (took only %.3f seconds)", w, duration.Seconds())
+		time.Sleep(time.Duration(w) * time.Second)
+	}
+	// Promote replica to new master.
+	promoteSlaveQuery := "STOP SLAVE;RESET SLAVE ALL;SET GLOBAL read_only = OFF;"
+	replica.VttabletProcess.QueryTablet(promoteSlaveQuery, keyspaceUnshardedName, true)
+
+	if replica.VttabletProcess.EnableSemiSync {
+		replica.VttabletProcess.QueryTablet("SET GLOBAL rpl_semi_sync_master_enabled = 1", keyspaceUnshardedName, true)
+	}
+	oldMaster := master
+	newMaster := replica
+
+	// Configure old master to use new master.
+	qr, err := newMaster.VttabletProcess.QueryTablet("SELECT @@GLOBAL.gtid_executed", keyspaceUnshardedName, true)
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+	val := qr.Rows[0][0]
+	gtID := val.String()
+	newPos := "MySQL56/" + gtID
+	fmt.Printf("New master position: %s", newPos)
+
+	// Use 'localhost' as hostname because Travis CI worker hostnames
+	// are too long for MySQL replication.
+	changeMasterCommands := fmt.Sprintf("RESET SLAVE;SET GLOBAL gtid_slave_pos = '%s';CHANGE MASTER TO MASTER_HOST='%s', MASTER_PORT=%d ,MASTER_USER='vt_repl', MASTER_USE_GTID = slave_pos;START SLAVE;", gtID, "localhost", newMaster.MySQLPort)
+	oldMaster.VttabletProcess.QueryTablet(changeMasterCommands, keyspaceUnshardedName, true)
+
+	// Notify the new vttablet master about the reparent.
+	clusterInstance.VtctlclientProcess.ExecuteCommand("TabletExternallyReparented", newMaster.Alias)
+}
+
+func waitForReplicationPos(t *testing.T, tabletA *cluster.Vttablet, tabletB *cluster.Vttablet, timeout float64) {
+	replicationPosA := getMasterPosition(t, tabletA)
+	for true {
+		replicationPosB := getMasterPosition(t, tabletB)
+		if positionAtLeast(t, tabletA, replicationPosB, replicationPosA) {
+			break
+		}
+		msg := fmt.Sprintf("%s's replication position to catch up %s's;currently at: %s, waiting to catch up to: %s", tabletB.Alias, tabletA.Alias, replicationPosB, replicationPosA)
+		waitStep(t, msg, timeout, 0.01)
+	}
+}
+
+func getMasterPosition(t *testing.T, tablet *cluster.Vttablet) string {
+	masterPositionQuery := "SELECT @@GLOBAL.gtid_executed;"
+	qr, err := tablet.VttabletProcess.QueryTablet(masterPositionQuery, keyspaceUnshardedName, true)
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+	val := qr.Rows[0][0]
+	gtID := val.ToString()
+	newPos := "MySQL56/" + gtID
+	return newPos
+}
```

Comment:
You could also do this by using grpc client.  `Tabletmanagerclient.MasterPosition`
Ref: in go/test/endtoend/tabletmanager/main_test.go

@arindamnayak We should look into moving these methods to a Util file.

Reply:
Good points.

Reply:
done